academic-observatory-workflows/academic_observatory_workflows/pubmed_telescope/telescope.py

```diff
@@ -0,0 +1,539 @@
+
+from observatory_platform.airflow.airflow import on_failure_callback
+from observatory_platform.airflow.sensors import PreviousDagRunSensor
+from observatory_platform.airflow.workflow import CloudWorkspace
+from observatory_platform.airflow.tasks import check_dependencies, gke_create_storage, gke_delete_storage
+from observatory_platform.google.gke import GkeParams, gke_make_kubernetes_task_params, gke_make_container_resources
+
+
+class DagParams:
+    """
+    :param dag_id: the id of the DAG.
+    :param cloud_workspace: Cloud settings.
+    :param bq_dataset_id: Dataset name for final tables.
+    :param api_bq_dataset_id: The dataset ID of the bigquery API.
+    :param bq_main_table_name: Table name of the final Pubmed table.
+    :param bq_upsert_table_name: Table name of the Pubmed upsert table.
+    :param bq_delete_table_name: Table name of the Pubmed delete table.
+    :param bq_dataset_description: Description of the Pubmed dataset.
+    :param start_date: The start date of the DAG.
+    :param schedule: How often the DAG should run.
+    :param ftp_server_url: Server address of Pubmed's FTP server.
+    :param ftp_port: Port for connectiong to Pubmed's FTP server.
+    :param reset_ftp_counter: Resets FTP connection after downloading x number of files.
+    :param max_download_attempt: Maximum number of download attempts of a single Pubmed file from the FTP server before throwing an error.
+    :param snapshot_expiry_days: How long until the backup snapshot (before this release's upserts and deletes) of the Pubmed table exist in BQ.
+    :param max_processes: Max number of parallel processes.
+    :param max_active_runs: the maximum number of DAG runs that can be run at once.
+    :param retries: the number of times to retry a task.
+    :param test_run: Whether this is a test run or not.
+    :param gke_namespace: The cluster namespace to use.
+    :param gke_volume_name: The name of the persistent volume to create
+    :param gke_volume_size: The amount of storage to request for the persistent volume in GiB
+    :param kwargs: Takes kwargs for building a GkeParams object.
+    """
+
+    def __init__(
+        self,
+        dag_id: str,
+        cloud_workspace: CloudWorkspace,
+        bq_dataset_id: str = "pubmed",
+        api_bq_dataset_id: str = "pubmed",
+        bq_main_table_name: str = "pubmed",
+        bq_upsert_table_name: str = "pubmed_upsert",
+        bq_delete_table_name: str = "pubmed_delete",
+        bq_dataset_description: str = "Pubmed Medline database, only PubmedArticle records: https://pubmed.ncbi.nlm.nih.gov/about/",
+        start_date: pendulum.DateTime = pendulum.datetime(year=2021, month=1, day=1),
+        schedule: str = "@weekly",
+        ftp_server_url: str = "ftp.ncbi.nlm.nih.gov",
+        ftp_port: int = 21,
+        reset_ftp_counter: int = 40,
+        max_download_attempt: int = 5,
+        snapshot_expiry_days: int = 31,
+        max_processes: int = 4,  # Limited to 4 due to RAM usage.
```

Comment:
May not be necessary anymore now that we are using Kubernetes and the transform task will have all of the RAM to itself. If not enough then we can increase the RAM and keep the CPUs the same.

I would then remove this parameter and us os.cpu_count() in this function in tasks.py:

```python
def updatefiles_merge_upserts_deletes(release: dict, updatefiles: List[Dict], max_processes: int = os.cpu_count()) -> None:
```

We don't want to use os.cpu_count() in the DAG, as it will then set the CPU count based on the Astro pod where the DAG is loaded I think.

We also need to remove / fix this same thing with Crossref Metadata, as it is setting os.cpu_count in the DagParams too.

Reply:
Oh good catch! Yeah i believe cpu_count() will be done at dag creation time on the astro machine. Should be fixable by simply putting it in the task itself which will be done at runtime.

Reply:
Yeah that works, thanks for fixing it.